cborg/src/Codec/CBOR/Decoding.hs

```diff
@@ -315,38 +315,46 @@ getDecodeAction (Decoder k) = k (\x -> return (Done x))
 toInt8   :: Int# -> Int8
 toInt16  :: Int# -> Int16
 toInt32  :: Int# -> Int32
-toInt64  :: Int# -> Int64
 toWord8  :: Word# -> Word8
 toWord16 :: Word# -> Word16
 toWord32 :: Word# -> Word32
+#if defined(ARCH_64bit)
+toInt64  :: Int# -> Int64
 toWord64 :: Word# -> Word64
+#else
+toInt64  :: Int64# -> Int64
+toWord64 :: Word64# -> Word64
+#endif
 #if MIN_VERSION_ghc_prim(0,8,0)
 toInt8   n = I8#  (intToInt8# n)
 toInt16  n = I16# (intToInt16# n)
 toInt32  n = I32# (intToInt32# n)
 toWord8  n = W8#  (wordToWord8# n)
 toWord16 n = W16# (wordToWord16# n)
 toWord32 n = W32# (wordToWord32# n)
-#if WORD_SIZE_IN_BITS == 64
+#if defined(ARCH_64bit)
 #if MIN_VERSION_base(4,17,0)
 toInt64  n = I64# (intToInt64# n)
 toWord64 n = W64# (wordToWord64# n)
 #else
 toInt64  n = I64# n
 toWord64 n = W64# n
 #endif
-#else
-toInt64  n = I64# (intToInt64# n)
-toWord64 n = W64# (wordToWord64# n)
 #endif
 #else
 toInt8   n = I8#  n
 toInt16  n = I16# n
 toInt32  n = I32# n
-toInt64  n = I64# n
 toWord8  n = W8#  n
 toWord16 n = W16# n
 toWord32 n = W32# n
+#if defined(ARCH_64bit)
+toInt64  n = I64# n
+toWord64 n = W64# n
+#endif
+#endif
+#if defined(ARCH_32bit)
+toInt64  n = I64# n
```

Comment:
It's unclear why this identity change is needed. The definitions in the 64 and 32bit cases here are the same as each other, and they are the same definitions as what where here before this code change.

I was half expecting it to be an identity refactoring in preparation for the next patch, but it's not. Was this intended to be something else? Otherwise it has no effect but makes the logic more complicated.

Reply:
> they are the same definitions as what where here before this code change.

No, they are different for 32bit and GHC >=9.4: Before this change, this meant that we compiled to `toInt64 n = I64# (intToInt64# n)` (incorrect, see https://github.com/well-typed/cborg/pull/322#discussion_r1325736908), and with this change, to `toInt64 n = I64# n` (correct).

But I just realized that this diff belongs into the second, not the first commit; will change that and also try to simplify the logic a bit.